BUG 1400887 Reorganize the Out of Resource Handling Topic

mburke5678 opened the pull request.

sjenning commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -155,111 +346,146 @@ threshold must hold before triggering a pod eviction.
 If an eviction criteria is satisfied, the node initiates the process of
 reclaiming the pressured resource until it observes that the signal has gone
 below its defined threshold. During this time, the node does not support
 scheduling any new pods.
```
Ah, I see this is mentioned here.

sjenning commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -270,21 +496,37 @@ xref:../admin_guide/overcommit.adoc#qos-classes[quality of service], and, among
 first.
-* `Burstable`: pods that consume the most of the starved resource relative to their
+
+|`Burstable`
```
I'd move Burstable above BestEffort in the table to order from least-to-most likely to be evicted i.e. highest-to-lowest QoS.

sjenning commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -298,7 +540,7 @@ those pods first.
 [NOTE]
 ====
-At this time, volumes that are backed by local disk are only deleted when a pod
+Volumes that are backed by local disk are only deleted when a pod
```
> Volumes that are backed by local disk are only deleted when a pod is deleted from the API server instead of when the pod is terminated.

This is not true as of 1.6.  They are deleted when the pod is terminated.

Probably just need to drop this whole note section now.

mburke5678 replied:
When we rev Kubernetes, are the changes implemented in earlier versions of OpenShift, 3.5 and 3.4?

mburke5678 commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -99,51 +240,101 @@ deprecated in favor of a pure disk eviction based configuration.
+[NOTE]
+====
+The node evaluates and monitors eviction thresholds every 10 seconds and the
+value can not be modified. This is the housekeeping interval.
```
Is this threshold the METRICS_RESOLUTION setting discussed in the release notes as having been changed in 3.5?
https://docs.openshift.com/container-platform/3.5/release_notes/ocp_3_5_release_notes.html#ocp-35-scale 

If so, would this change have been applied back to 3.4 and 3.3?

mburke5678 replied:
Added: "The cAdvisor housekeeping interval is 30 seconds by default."
The node auto-discovers these file systems using `cAdvisor`. The cAdvisor housekeeping interval is 30 seconds by default.

derekwaynecarr replied:
please do not add anything about cadvisor.  just state that we monitor every 10s.

mburke5678 replied:
Removed cAdvisor. 
"{product-title} monitors these file systems every 10 seconds. "

mburke5678 commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,160 @@ toc::[]
+* For a hard threshold, enter:
++
+----
+kubelet ---eviction-hard="<eviction_signal><operator><quantity>"
```
Hidden for the moment

derekwaynecarr commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,161 @@ toc::[]
++
+For example:
++
+.Sample Node Configuration file for a hard eviction 
```
i am a little worried that people will assume this is recommended configuration.

mburke5678 replied:
Added the bolded phrase (not bold in doc):
Add or edit the eviction thresholds under the `eviction-hard` or `eviction-soft` paramters, **using the following examples as general guidelines and not recommended settings.**

derekwaynecarr commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,161 @@ toc::[]
+<2> An eviction threshold based on a specific eviction trigger signal.
+====
++
+.Sample Node Configuration file for a soft eviction
```
since we discussed soft eviction as being rather advanced, i am also worried people will take this as preferred config.

mburke5678 replied:
Added the bolded phrase (not bold in doc):
Add or edit the eviction thresholds under the eviction-hard or eviction-soft paramters, **using the following examples as general guidelines and not recommended settings.**

derekwaynecarr commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,161 @@ toc::[]
-*_node-config.yaml_* file.
+//
+[[out-of-resource-create-cli]]
+=== Using the Node CLI to Create a Policy
```
is the comment hiding this?  we don't have a node cli so this section doesnt make much sense in openshift context.

mburke5678 replied:
Removed the section

derekwaynecarr commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -99,51 +239,104 @@ deprecated in favor of a pure disk eviction based configuration.
 administrator-specified grace period. The node does not reclaim resources
 associated with the eviction signal until that grace period is exceeded. If no
-grace period is provided, the node errors on startup.
+grace period is provided in the node config the node errors on startup or the CLI command will fail.
```
remove or the CLI command will fail.

mburke5678 replied:
Removed

mburke5678 commented:
@derekwaynecarr Made changes to address your concerns. PTAL.

mburke5678 commented:
@derekwaynecarr Made changes to address your concerns. PTAL.

mburke5678 commented:
@sjenning @derekwaynecarr This is the PR I mentioned at scrum this morning. Want to make sure you are satisfied that I addressed your concerns before moving to QE.

sjenning commented:
thanks @mburke5678 lgtm

mburke5678 commented:
@qwang1  Please review so that I can move forward with these changes.

mburke5678 commented:
https://trello.com/c/MKBBB6JQ/709-1-improve-documentation-for-out-of-resource-handling

qwang1 commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -53,17 +139,26 @@ curl <certificate details> \
+|`*MemoryPressure*`
 |`*memory.available*`
 |`memory.available` = `node.status.capacity[memory]` - `node.stats.memory.workingSet`
+| Available memory on the node has eceeded an eviction threshold
```
Typo: eceeded

qwang1 replied:
Typo: eceeded

qwang1 commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,14 +170,23 @@ curl <certificate details> \
-The node supports two file system partitions when detecting disk pressure.
+Each of the above signals supports either a literal or percentage-based value. The percentage-based value is calculated relative to the total capacity associated with each signal.
+
+A script derives the value for `memory.available` from the `cgroupfs` driver using the same set of steps that the kubelet performs. The script excludes *inactive_file* memory (that is, the number of bytes of file-backed memory on inactive LRU list) from its calculation as it assumes that *inactive_file* memory is reclaimable under pressure. 
```
It's not clear to me that "cgroupfs" driver here is correct or not. Generally "--cgroup-driver=systemd" is configured.

mburke5678 replied:
@qwang1 Changed this to: "from your `cgroup` driver." Does this make more sense?

qwang1 replied:
@mburke5678 Make sense.

qwang1 commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -387,14 +587,14 @@ To reach that capacity, either some pod is using more than its request, or the
 
 ----
 capacity = 10 Gi
 system-reserved = 10 Gi * .01 = 1 Gi
```
".01" this style of writing is strange to me, an non-native speaker~

mburke5678 replied:
@qwang1 .01 is a  decimal;  it is equal to 1/100th.
"10 Gi * .01 = 1 Gi" is a mathematical equation.

qwang1 replied:
@mburke5678 It's my first time to see this format. A new thing for me, thanks. Since it's a universal writing, I can accept it.

mburke5678 commented:
@openshift/team-documentation PTAL

mburke5678 commented:
@openshift/team-documentation  BUMP

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
 
+Administrators can proactively monitor nodes for and prevent against situations where the node runs out of compute and memory resources using configurable xref:out-of-resource-eviction-policy[eviction policies]. 
+
+This topic also provides information on how {product-title} handles out-of-resource conditions and provides an xref:out-of-resource-schedulable-resources-and-eviction-policies[example scenario] and xref:out-of-resource-recommended-practices[best practices]:
```
s/best/recommended

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
+Failing a pod allows the node to reclaim needed resources.  
+
+An eviciton policy is a combination of an xref:out-of-resource-eviction-signals[eviction trigger signal] with a specific xref:out-of-resource-eviction-thresholds[eviction threshold value], that is set
+in the xref:out-of-resource-eviction-policy[node configuration file] or through the xref:out-of-resource-eviction-policy-cli[node command line]. Evictions can be either hard, where a node takes immediate action on a pod that exceeds a threshold, or soft, where a node allows a grace period before taking action. Review the sections below for important information the differences between xref:out-of ...
```
What's the "node command line" linking to? That xref doesn't lead anywhere and I can't tell from the bit inside the square brackets. Also s/Review/See.

mburke5678 replied:
Should be "command line." Fixed link.

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
-fail one or more pods in order to reclaim the starved resource using an eviction
-policy. When the node fails a pod, it terminates all containers in the pod, and
-the `PodPhase` is transitioned to *Failed*.
+. Add or edit the eviction thresholds under the `eviction-hard` or `eviction-soft` paramters, using the following examples
```
s/paramters/parameters

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
 
-Using eviction policies, a node can proactively monitor for and prevent
-against total starvation of a compute resource.
+. Edit the node configuration file on a node.
```
I think the above step can be merged with the step below. Something like:

"Edit the edit the eviction thresholds under the `eviction-hard` or `eviction-soft` parameters in a node configuration file:"

mburke5678 replied:
Removed the numbered list and combined the steps into:
"To configure an eviction policy using the node configuration file Edit the eviction thresholds under the eviction-hard or eviction-soft parameters in a node configuration file."

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
+
+<1> The type of eviction: Use this parameter for a xref:out-of-resource-hard-eviction-thresholds[soft eviction].
+<2> An eviction threshold based on a specific eviction trigger signal.
+<3> The grace period for the soft eviction. Leave the default values for best performance.
```
s/best/ideal

?

mburke5678 replied:
Changed to "optimal performance." Search in Git found a couple "optimal performance" and one "best... performance", but no "ideal performance."

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -12,39 +12,120 @@ toc::[]
+<3> The grace period for the soft eviction. Leave the default values for best performance.
+====
+
+. Restart the {product-title} services for the changes to take effect:
```
It's saying "services" but we're only restarting one service. Should more be added? Should it be "service"?

mburke5678 replied:
Changed to "service"

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
-The node supports two file system partitions when detecting disk pressure.
+Each of the above signals supports either a literal or percentage-based value. The percentage-based value is calculated relative to the total capacity associated with each signal.
+
+A script derives the value for `memory.available` from your cgroup driver using the same set of steps that the kubelet performs. The script excludes *inactive_file* memory (that is, the number of bytes of file-backed memory on inactive LRU list) from its calculation as it assumes that *inactive_file* memory is reclaimable under pressure. 
```
I'm not sure what *inactive_file* is. If it's a parameter, maybe use backticks for markup?

mburke5678 replied:
inactive file is computer term, apparently. I changed to "inactive file".
https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Resource_Management_Guide/sec-memory.html

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
-=== Eviction Thresholds
+=== Understanding Eviction Thresholds
 
 You can configure a node to specify eviction thresholds, which trigger the node
```
s/trigger/triggers

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
 
-To protect this, set the following flag to control how long the node must wait
-before transitioning out of a pressure condition:
+A soft eviction threshold pairs an eviction threshold with a required administrator-specified grace period. The node does not reclaim resources associated with the eviction signal until that grace period is exceeded. If no grace period is provided in the node config the node errors on startup.
```
s/config/configuration file

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
+
+. Edit the node configuration file (the *_/etc/origin/node/node-config.yaml_* file).
+
+. Add or modify the `system-reserved` parameter for `eviction-hard` or `eviction-soft`
```
Same comment as above. The two steps here could be merged.

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
+----
++
+The node will toggle the condition back to *false* when the node has not observed an eviction threshold being met
+for the specified pressure condition for the specified period.
```
Mainly because of the "will", I'd suggest:

"The node toggles back to *false* when it has not observed an eviction threshold being met for the specified pressure condition for the specified period."

bfallonf commented on admin_guide/out_of_resource_handling.adoc:
```diff
@@ -75,243 +165,353 @@ curl <certificate details> \
+The default choice is intended to allow the system to stabilize, and to prevent the scheduler from assigning new pods to the node before it has settled. 
+====
+
+. Restart the {product-title} services for the changes to take effect:
```
Same comment above with s/services/service

bfallonf commented:
@mburke5678 I made some comments. Let me know if there's anything confusing.

mburke5678 commented:
Made changes per Brice. As this is a reorg with minor clarifications no Rev History.

mburke5678 commented:
@sjenning @derekwaynecarr  Can the changes I made through this PR be applied back to OpenShift 3.3? Can I apply that one PR to update the topic in 3.3, 3.4, and 3.5?

derekwaynecarr commented:
@mburke5678 - it would need some tweaks.

3.3 only supported eviction on memory, and you could not use percentages.
3.4 added support for eviction on disk, and you could use percentages in thresholds.
